pkgs/applications/editors/tmcbeans/default.nix

```
@@ -0,0 +1,70 @@
+{ lib, stdenv, fetchurl, makeWrapper, makeDesktopItem, which, unzip,
+libicns, imagemagick, jdk, perl
+}:
+
+let
+  version = "1.4.0";
+  desktopItem = makeDesktopItem {
+    name = "tmcbeans";
+    exec = "tmcbeans";
+    comment = "Integrated Development Environment";
+    desktopName = "TMCBeans IDE";
+    genericName = "Integrated Development Environment";
+    categories = [ "Development" ];
+    icon = "tmcbeans";
+  };
+in
+stdenv.mkDerivation {
+  pname = "tmcbeans";
+  inherit version;
+  src = fetchurl {
+    url = "https://download.mooc.fi/tmcbeans/installers/tmcbeans_${version}.zip";
+    hash = "sha512-kugZ7DAarbKK8RvOx7VYIpCwNhm3QVBlvqy+R7oeDDTT0LeDtHsWNIpIQSgT7t9MExfG1X9bqCqspJwYQxUG5g==";
+  };
+
+  buildCommand = ''
+    # Unpack and perform some path patching.
+    unzip $src
+    patchShebangs .
+
+    rm tmcbeans/bin/*.exe
+
+    # Copy to installation directory and create a wrapper capable of starting
+    # it.
+    mkdir -pv $out/bin
+    cp -a tmcbeans $out
+    makeWrapper $out/tmcbeans/bin/tmcbeans $out/bin/tmcbeans\
+      --prefix PATH : ${lib.makeBinPath [ jdk which ]} \
+      --prefix JAVA_HOME : ${jdk.home} \
+      --add-flags "--jdkhome ${jdk.home} \
+      -J-Dawt.useSystemAAFontSettings=on -J-Dswing.aatext=true"
+
+    # Extract pngs from the Apple icon image and create
+    # the missing ones from the 1024x1024 image.
+    icns2png --extract $out/tmcbeans/nb/tmcbeans.icns
+    for size in 16 24 32 48 64 128 256 512 1024; do
+      mkdir -pv $out/share/icons/hicolor/"$size"x"$size"/apps
+      if [ -e tmcbeans_"$size"x"$size"x32.png ]
+      then
+        mv tmcbeans_"$size"x"$size"x32.png $out/share/icons/hicolor/"$size"x"$size"/apps/tmcbeans.png
+      else
+        convert -resize "$size"x"$size" $out/tmcbeans/nb/tmcbeans.png $out/share/icons/hicolor/"$size"x"$size"/apps/tmcbeans.png
+      fi
+    done;
+
+    # Create desktop item, so we can pick it from the KDE/GNOME menu
+    mkdir -pv $out/share/applications
+    ln -s ${desktopItem}/share/applications/* $out/share/applications
+  '';
+
+  nativeBuildInputs = [ makeWrapper unzip ];
+  buildInputs = [ perl libicns imagemagick ];
+
+  meta = {
+    description = "A customized version of the Netbeans IDE for Java, C, C++ and PHP with the Test My Code system installed";
+    homepage = "https://www.mooc.fi";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ robbins ];
+    platforms = lib.platforms.unix;
```

This is the default. Since source is binary, we also need to add this.
```suggestion
    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
```